### DISCUSSION TITLE
Mechanism for OS-specific configuration/keybindings?

### DISCUSSION THREAD

**Discussion by 0xzhzh** [ghostty-org/ghostty#2865] [Q&A] [answered]
I'm trying to build out my config for both macOS and Linux system. I am hoping to reuse as much of the same config as possible, but some settings---key bindings in particular---are OS-specific. I like to use `super+` on macOS, to fit in with what other native appliations do, and `ctrl+shift+` on Linux, where that is the norm (which leaves me the `super+` "namespace" for my window manager).

I'm wondering whether there is any existing mechanism for setting up some bindings for only one OS, and some for only another?

The only mechanism I can think of at the moment is "fallible inclusion", i.e. something like:
```
config-file = ?macos-config
config-file = ?linux-config
```

However, I don't like that it keeps two sets of key bindings in two different files, because that makes them more difficult to maintain and keep consistent between systems. It also makes bootstrapping a new machine slightly more complicated.

**Comment by jcollie** [comment]
"Fallible inclusion" was merged in #2243 .

See discussion #2356 for an exhaustive discussion of a conditional config extension.

A partial solution, at least for themes, was implemented in #2735 .

**Reply by 0xzhzh** [reply]
The partial solution from #2735 applies only to themes and not to keybindings; it would be really nice to have something like:

```
keybind = macos:super+c=copy_to_clipboard
keybind = linux:ctrl+shift+c=copy_to_clipboard
```

The discussion in #2356 seems to have been fairly theme-specific, though some passing suggestions were more general-purpose (e.g., `keybind = something [os=linux]`, `if = os=linux`). But the discussion seems to have closed since #2725 was opened and #2735 was merged.

**Reply by mitchellh** [reply, marked as answer]
I agree that we don't have a good place to discuss this and I'm happy to make it this place.

As for what @jcollie said I think it's more helpful than it seems on first glance! He's right that #2735 applies a partial solution. The solution is only _exposed_ to themes but the _Ghostty internals_ were refactored to allow conditional configuration on _anything_ so the good news is the **difficult core work is done** (at least a big chunk of it).

What's leftover from that and #2356 is a bike shedding exercise on syntax. Once we determine syntax, we can take advantage of the system built in #2735 to enable it.

Additionally, you didn't mention it, but #2243 is very useful as a stopgap for this. Users in the community are using this as a way to do this by putting OS-specific configuration in files like `~/.config/ghostty/macos` and so on and then doing `config-file = ?macos` and that only works if the file exists. Then your system setup scripts (i.e. Nix) can only drop that file if the target OS matches. **This works today.**

**Comment by karlicoss** [comment]
A cheeky option if you want a mac-only override (e.g. I swap ctrl and super).
Name the macos mixin from lowercase:
```
ls .config/ghostty/
config.ghostty  macos
```
But include it starting with upper case:
```
config-file = ?Macos
```
That way it won't be loaded on linux (since it's case sensitive), but will be on macos.
